File: src-scala/org/jgll/recognizer/GLLRecognizer.scala
package org.jgll.recognizer

import org.jgll.util.InputTrait
import org.jgll.grammar.GrammarTrait
import org.jgll.grammar.slot.BodyGrammarSlotTrait
import org.jgll.grammar.slot.GrammarSlotTrait


trait GLLRecognizerTrait {
  self: GrammarTrait
   with BodyGrammarSlotTrait
   with GrammarSlotTrait
   with InputTrait
   with GSSNodeTrait
   with DescriptorTrait
  =>
  trait GLLRecognizer {

    def recognize(input: Input, grammar: Grammar, nonterminalName: String): Boolean

    def recognize(input: Input,
        start: Int,
        end: Rep[Int],
        slot: BodyGrammarSlot): Boolean

    def add(label: GrammarSlot, u: GSSNode, inputIndex: Int): Unit

    def pop(u: GSSNode, i: Int): Unit

    def create(L: GrammarSlot, u: GSSNode, i: Int): GSSNode

    def hasNextDescriptor(): Boolean

    def nextDescriptor(): Descriptor

    def getCi(): Int

    def getCu(): GSSNode

    def update(gssNode: GSSNode, inputIndex: Int): Unit

    def recognitionError(gssNode: GSSNode, inputIndex: Int): Unit
  }
}

File: src/org/jgll/recognizer/GLLRecognizer.java
package org.jgll.recognizer;

import org.jgll.grammar.Grammar;
import org.jgll.grammar.slot.BodyGrammarSlot;
import org.jgll.grammar.slot.GrammarSlot;
import org.jgll.util.Input;

public interface GLLRecognizer {

	public boolean recognize(Input input, Grammar grammar, String nonterminalName);
	
	public boolean recognize(Input input, int start, int end, BodyGrammarSlot slot);
	
	public void add(GrammarSlot label, GSSNode u, int inputIndex);

	public void pop(GSSNode u, int i);

	public GSSNode create(GrammarSlot L, GSSNode u, int i);
	
	public boolean hasNextDescriptor();
	
	public Descriptor nextDescriptor();
	
	public int getCi();
	
	public GSSNode getCu();
	
	public void update(GSSNode gssNode, int inputIndex);
	
	public void recognitionError(GSSNode gssNode, int inputIndex);
	
}
